### DISCUSSION TITLE
Scoped critical exceptions

### DISCUSSION THREAD

**Discussion by stijnmoreels** [arcus-azure/arcus.security#283] [Ideas]
Right now, when we add an critical exception as failure that should halt the secret retrieval (instead of going to the next secret provider), this critical exception is added for _all_ registered secret providers.
Meaning, when secret provider A fails with a critical exception type that secret provider B has registered, the secret store will still halt the secret retrieval process.

Should we introduce scopes for this? As in: critical exceptions that were registered for secret provider A only works with secret provider A instead of _all_ secret providers?

**Comment by stijnmoreels** [comment]
Thoughts? @tomkerkhove , @mbraekman , @fgheysels ?

**Reply by fgheysels** [reply]
Can you elaborate ?  Can you give an example of a critical exception ?  In which situations do you want to halt the secret retrieval when a provider throws an exception ?

**Reply by stijnmoreels** [reply]
Currently we already support critical exceptions. They are specific exceptions that the secret store handles differently. (see docs for more info on this, https://security.arcus-azure.net/features/secret-store/create-new-secret-provider#adding-critical-exceptions)
The question here is to stricten the scope where this type of exception is considered critical.

An example of such an exception is the `AuthenticationException`. The authentication process with an external system is considered critical so it seems logical when your secret provider throws this, the secret retrieval will be halted (read: no other secret providers will be tried as a throwal of an critical exception is considered fatal and usually the result of misconfiguration).

Now, what I'm proposing is that we scope these critical exceptions. As in: when another secret provider also throws an `AuthenticationException` it has to be registered for that secret provider first, instead of using all registered critical exceptions.

**Reply by tomkerkhove** [reply]
Personally I would keep it as is until there is demand for it - What do you think?

**Reply by stijnmoreels** [reply]
@tomkerkhove , yeah, is ok of course 👍 , was thinking of possible scenario's where this could become an issue or when it's failing for the wrong reasons. But maybe I'm too paranoid 😄 .
